stan/math/prim/fun/LDLT_factor.hpp

```diff
@@ -12,120 +12,69 @@ namespace stan {
- *
- * The caller can use the LDLT_factor objects as needed.  For
- * instance
- *
- * ~~~
- * x1 = mdivide_left_ldlt(ldlt_A1, b1);
- * x2 = mdivide_right_ldlt(b2, ldlt_A2);
- *
- * d1 = log_determinant_ldlt(ldlt_A1);
- * d2 = log_determinant_ldlt(ldlt_A2);
- * ~~~
- *
- * This class is conceptually similar to the corresponding Eigen
- * class.  Any symmetric, positive-definite matrix A can be
- * decomposed as LDL' where L is unit lower-triangular and D is
- * diagonal with positive diagonal elements.
+ * LDLT_factors are cheap to copy (they copy references to the
+ * underlying storage).
  *
  * @tparam T type of elements in the matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam alloc_in_arena allocate members of LDLT_factor in the arena
  */
-template <typename T, int R, int C>
-class LDLT_factor {
- public:
-  using vector_t = Eigen::Matrix<T, Eigen::Dynamic, 1>;
-  using matrix_t = Eigen::Matrix<T, R, C>;
-  using ldlt_t = Eigen::LDLT<matrix_t>;
-  using size_type = size_t;
-  using value_type = double;
-
-  LDLT_factor() : N_(0), ldltP_(new ldlt_t()) {}
-
-  explicit LDLT_factor(const matrix_t& A) : N_(0), ldltP_(new ldlt_t()) {
-    compute(A);
-  }
-
-  inline void compute(const matrix_t& A) {
-    check_square("LDLT_factor", "A", A);
-    N_ = A.rows();
-    ldltP_->compute(A);
-  }
-
-  inline bool success() const {
-    if (ldltP_->info() != Eigen::Success) {
-      return false;
-    }
-    if (!(ldltP_->isPositive())) {
-      return false;
-    }
-    vector_t ldltP_diag(ldltP_->vectorD());
-    for (int i = 0; i < ldltP_diag.size(); ++i) {
-      if (ldltP_diag(i) <= 0 || is_nan(ldltP_diag(i))) {
-        return false;
-      }
-    }
-    return true;
-  }
-
-  inline T log_abs_det() const { return sum(log(ldltP_->vectorD())); }
+template <typename T, bool alloc_in_arena>
+class LDLT_factor;
 
-  inline void inverse(matrix_t& invA) const {
-    invA.setIdentity(N_);
-    ldltP_->solveInPlace(invA);
-  }
-
-  template <typename Rhs>
-  inline const Eigen::Solve<ldlt_t, Rhs> solve(
-      const Eigen::MatrixBase<Rhs>& b) const {
-    return ldltP_->solve(b);
-  }
-
-  inline matrix_t solveRight(const matrix_t& B) const {
-    return ldltP_->solve(B.transpose()).transpose();
-  }
-
-  inline vector_t vectorD() const { return ldltP_->vectorD(); }
-
-  inline ldlt_t matrixLDLT() const { return ldltP_->matrixLDLT(); }
-
-  inline size_t rows() const { return N_; }
-  inline size_t cols() const { return N_; }
+/**
+ * An LDLT_factor with `alloc_in_arena = False` is a structure that
+ * holds a matrix of type T and the LDLT of its values.
+ *
+ * @tparam T type of elements in the matrix
+ */
+template <typename T>
+class LDLT_factor<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>, false> {
+ private:
+  using ldlt_type
+      = Eigen::LDLT<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>;
+  std::shared_ptr<ldlt_type> ldlt_ptr_;
```

Comment:
I'm not following why this needs to be a shared pointer (or pointer at all?)

Reply:
1. Makes copies of this cheap (it gets copied into `reverse_pass_callback`)
2. Handles the deletin' for me

Reply:
Shared pointers are pretty expensive. Each one essentially has a control block that keeps track of the count. Could we use unique_ptr here?

I think I need to read more about how this type is used to have a better grasp of this

Reply:
I ditched the shared pointer and made LDLT_factors heavy weight objects (that shouldn't be copied)